File: packages/capacitor-actito-user-inbox/ios/Sources/ActitoUserInboxPlugin/ActitoUserInboxPlugin.swift
import ActitoKit
import ActitoUserInboxKit
import Capacitor
import Foundation

@objc(ActitoUserInboxPlugin)
public class ActitoUserInboxPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "ActitoUserInboxPlugin"
    public let jsName = "ActitoUserInboxPlugin"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "parseResponseFromJson", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "parseResponseFromString", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "open", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "markAsRead", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "remove", returnType: CAPPluginReturnPromise),
    ]

    @objc func parseResponseFromJson(_ call: CAPPluginCall) {
        guard let json = call.getObject("json") else {
            call.reject("Missing 'json' parameter.")
            return
        }

        do {
            let response = try Actito.shared.userInbox().parseResponse(json: json)

            call.resolve([
                "result": try response.toJson()
            ])
        } catch {
            call.reject(error.localizedDescription)
        }
    }

    @objc func parseResponseFromString(_ call: CAPPluginCall) {
        guard let json = call.getString("json") else {
            call.reject("Missing 'json' parameter.")
            return
        }

        do {
            let response = try Actito.shared.userInbox().parseResponse(string: json)

            call.resolve([
                "result": try response.toJson()
            ])
        } catch {
            call.reject(error.localizedDescription)
        }
    }

    @objc func open(_ call: CAPPluginCall) {
        guard let json = call.getObject("item") else {
            call.reject("Missing 'item' parameter.")
            return
        }

        let item: ActitoUserInboxItem

        do {
            item = try ActitoUserInboxItem.fromJson(json: json)
        } catch {
            call.reject(error.localizedDescription)
            return
        }

        Actito.shared.userInbox().open(item) { result in
            switch result {
            case let .success(notification):
                do {
                    call.resolve([
                        "result": try notification.toJson()
                    ])
                } catch {
                    call.reject(error.localizedDescription)
                }
            case let .failure(error):
                call.reject(error.localizedDescription)
            }
        }
    }

    @objc func markAsRead(_ call: CAPPluginCall) {
        guard let json = call.getObject("item") else {
            call.reject("Missing 'item' parameter.")
            return
        }

        let item: ActitoUserInboxItem

        do {
            item = try ActitoUserInboxItem.fromJson(json: json)
        } catch {
            call.reject(error.localizedDescription)
            return
        }

        Actito.shared.userInbox().markAsRead(item) { result in
            switch result {
            case .success:
                call.resolve()
            case let .failure(error):
                call.reject(error.localizedDescription)
            }
        }
    }

    @objc func remove(_ call: CAPPluginCall) {
        guard let json = call.getObject("item") else {
            call.reject("Missing 'item' parameter.")
            return
        }

        let item: ActitoUserInboxItem

        do {
            item = try ActitoUserInboxItem.fromJson(json: json)
        } catch {
            call.reject(error.localizedDescription)
            return
        }

        Actito.shared.userInbox().remove(item) { result in
            switch result {
            case .success:
                call.resolve()
            case let .failure(error):
                call.reject(error.localizedDescription)
            }
        }
    }
}

File: packages/capacitor-actito-user-inbox/android/src/main/java/com/actito/inbox/user/capacitor/ActitoUserInboxPlugin.kt
package com.actito.inbox.user.capacitor

import com.actito.Actito
import com.actito.ActitoCallback
import com.actito.inbox.user.ktx.userInbox
import com.actito.inbox.user.models.ActitoUserInboxItem
import com.actito.models.ActitoNotification
import com.getcapacitor.JSObject
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.CapacitorPlugin


@CapacitorPlugin(name = "ActitoUserInboxPlugin")
public class ActitoUserInboxPlugin : Plugin() {
    @PluginMethod
    public fun parseResponseFromJson(call: PluginCall) {
        val json = call.getObject("json") ?: run {
            call.reject("Missing 'json' parameter.")
            return
        }

        try {
            val response = Actito.userInbox().parseResponse(json)
            call.resolve(
                JSObject().apply {
                    put("result", response.toJson())
                }
            )
        } catch (e: Exception) {
            call.reject(e.localizedMessage)
        }
    }

    @PluginMethod
    public fun parseResponseFromString(call: PluginCall) {
        val json = call.getString("json") ?: run {
            call.reject("Missing 'json' parameter.")
            return
        }

        try {
            val response = Actito.userInbox().parseResponse(json)
            call.resolve(
                JSObject().apply {
                    put("result", response.toJson())
                }
            )
        } catch (e: Exception) {
            call.reject(e.localizedMessage)
        }
    }

    @PluginMethod
    public fun open(call: PluginCall) {
        val json = call.getObject("item") ?: run {
            call.reject("Missing 'item' parameter.")
            return
        }

        val item: ActitoUserInboxItem = try {
            ActitoUserInboxItem.fromJson(json)
        } catch (e: Exception) {
            call.reject(e.localizedMessage)
            return
        }

        Actito.userInbox().open(item, object : ActitoCallback<ActitoNotification> {
            override fun onSuccess(result: ActitoNotification) {
                try {
                    call.resolve(
                        JSObject().apply {
                            put("result", result.toJson())
                        }
                    )
                } catch (e: Exception) {
                    call.reject(e.localizedMessage)
                }
            }

            override fun onFailure(e: Exception) {
                call.reject(e.localizedMessage)
            }
        })
    }

    @PluginMethod
    public fun markAsRead(call: PluginCall) {
        val json = call.getObject("item") ?: run {
            call.reject("Missing 'item' parameter.")
            return
        }

        val item: ActitoUserInboxItem = try {
            ActitoUserInboxItem.fromJson(json)
        } catch (e: Exception) {
            call.reject(e.localizedMessage)
            return
        }

        Actito.userInbox().markAsRead(item, object : ActitoCallback<Unit> {
            override fun onSuccess(result: Unit) {
                call.resolve()
            }

            override fun onFailure(e: Exception) {
                call.reject(e.localizedMessage)
            }
        })
    }

    @PluginMethod
    public fun remove(call: PluginCall) {
        val json = call.getObject("item") ?: run {
            call.reject("Missing 'item' parameter.")
            return
        }

        val item: ActitoUserInboxItem = try {
            ActitoUserInboxItem.fromJson(json)
        } catch (e: Exception) {
            call.reject(e.localizedMessage)
            return
        }

        Actito.userInbox().remove(item, object : ActitoCallback<Unit> {
            override fun onSuccess(result: Unit) {
                call.resolve()
            }

            override fun onFailure(e: Exception) {
                call.reject(e.localizedMessage)
            }
        })
    }
}
